### PR TITLE
Cancel no longer needed HTTP requests from PM UI when package list changes

### LINKED ISSUE #10801: PM UI should cancel HTTP requests when package list changes
In the Package Manager UI, there are a number of scenarios where NuGet might start HTTP requests, but not be interested in the result before they return:

* Changing tabs
* Changing search query text before the previous result comes back

By cancelling the requests, we reduce the risk that a customer's machine is going to be busy completing unnecessary HTTP requests, which could lead to perf problems for necessary HTTP requests.

related, but not duplicates:
https://github.com/NuGet/Home/issues/10814
https://github.com/NuGet/Home/issues/10815
https://github.com/NuGet/Home/issues/10816

### PR DESCRIPTION
**Pull request by zivkan**
## Bug

Fixes: https://github.com/NuGet/Home/issues/10801

Regression? Last working version:

## Description

* Add a cancellation token to the view model, passed to the search service
* Request cancellation on the token when the view model is no longer used
* Added new `NuGetSearchServiceReconnector` to abstract away the brokered service implementation details, and avoid multiple places needing to know about the brokered service, subscribe to the `AvailabilityChanged` event, unsubscribe from the event, keep a JTF just for responding to the availability change, etc.

## PR Checklist

- [X] PR has a meaningful title
- [X] PR has a linked issue.
- [X] Described changes

- **Tests**
  - [X] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [X] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs [hunk 1]
```diff
@@ -460,20 +465,30 @@ public BitmapSource IconBitmap
...
+            return await _searchService.GetPackageVersionsAsync(identity, Sources, IncludePrerelease, _cancellationTokenSource.Token);
+        }
 
-        public Lazy<Task<(PackageSearchMetadataContextInfo, PackageDeprecationMetadataContextInfo)>> DetailedPackageSearchMetadata { get; set; }
-        public Task<(PackageSearchMetadataContextInfo, PackageDeprecationMetadataContextInfo)> GetDetailedPackageSearchMetadataAsync() => (DetailedPackageSearchMetadata ?? LazyNullDetailedPackageSearchMetadata).Value;
+        // This Lazy/AsyncLazy is just because DetailControlModel calls GetDetailedPackageSearchMetadataAsync directly,
+        // and there are tests that don't mock IServiceBroker and INuGetSearchService. It's called via a jtf.RunAsync that is
+        // not awaited. By keeping this AsyncLazy, we ensure that the exception is thrown in an async continuation. Whereas
+        // if we get rid of it and have GetDetailedPackageSearchMetadataAsync call _searchService directly, then the exception
+        // will not be thrown in a continuation, and the test will fail.
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs:478]: I just noticed these comments are incomplete. Going to fix.

Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs:478]: Actually, it's not incomplete, I don't know what I was thinking 😕

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs [hunk 2]
```diff
@@ -619,42 +634,48 @@ private static void AddToCache(string cacheKey, BitmapSource iconBitmapImage)
...
-            LatestVersion = result;
-            Status = GetPackageStatus(LatestVersion, InstalledVersion, AutoReferenced);
-        }
-
         private async System.Threading.Tasks.Task ReloadPackageDeprecationAsync()
         {
-            PackageDeprecationMetadataContextInfo result = await _backgroundDeprecationMetadataLoader.Value;
+            try
+            {
+                var cancellationToken = _cancellationTokenSource.Token;
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs]: explicit type - or, I'd be okay without this variable at all.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+            var serviceBroker = new Mock<IServiceBroker>();
+            serviceBroker.Setup(sb =>
+#pragma warning disable ISB001 // Dispose of proxies
+            sb.GetProxyAsync<INuGetSearchService>(It.IsAny<ServiceJsonRpcDescriptor>(), It.IsAny<ServiceActivationOptions>(), It.IsAny<CancellationToken>())
+#pragma warning restore ISB001 // Dispose of proxies
+            )
+                .ReturnsAsync(searchService.Object);
+
+            var target = await NuGetSearchServiceReconnector.CreateAsync(serviceBroker.Object, _jtf, CancellationToken.None);
```
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs]: explicit type

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            // Act & Assert
+            ArgumentNullException exception = await Assert.ThrowsAsync<ArgumentNullException>(() => NuGetSearchServiceReconnector.CreateAsync(serviceBroker: null, _jtf, CancellationToken.None));
+            Assert.Equal("serviceBroker", exception.ParamName);
+
+            exception = await Assert.ThrowsAsync<ArgumentNullException>(() => NuGetSearchServiceReconnector.CreateAsync(serviceBroker.Object, jtf: null, CancellationToken.None));
+            Assert.Equal("jtf", exception.ParamName);
+        }
+
+        [Fact]
+        public async Task Object_NormalCreation_GetInstanceFromServiceBroker()
```
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs]: Consider a more specific name. This test is really about 1 call to `ContinueSearchAsync`, right?

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            int getProxyCount = 0;
+            var serviceBroker = new Mock<IServiceBroker>();
+            serviceBroker.Setup(sb =>
+#pragma warning disable ISB001 // Dispose of proxies
+            sb.GetProxyAsync<INuGetSearchService>(It.IsAny<ServiceJsonRpcDescriptor>(), It.IsAny<ServiceActivationOptions>(), It.IsAny<CancellationToken>())
+#pragma warning restore ISB001 // Dispose of proxies
+            )
+                .ReturnsAsync(() =>
+                {
+                    var count = Interlocked.Increment(ref getProxyCount);
```
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs]: explicit type

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+                    }
+                });
+
+            // Act
+            var target = await NuGetSearchServiceReconnector.CreateAsync(serviceBroker.Object, _jtf, CancellationToken.None);
+            await target.AvailabilityChangedAsync();
+            await target.Object.ContinueSearchAsync(CancellationToken.None);
+
+            // Assert
+            searchService2.Verify(s => s.ContinueSearchAsync(It.IsAny<CancellationToken>()), Times.Once);
```
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs]: nit: flip order (assert service1 first)

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+            var serviceBroker = new Mock<IServiceBroker>();
+            serviceBroker.Setup(sb =>
+#pragma warning disable ISB001 // Dispose of proxies
+            sb.GetProxyAsync<INuGetSearchService>(It.IsAny<ServiceJsonRpcDescriptor>(), It.IsAny<ServiceActivationOptions>(), It.IsAny<CancellationToken>())
+#pragma warning restore ISB001 // Dispose of proxies
+            )
+                .ReturnsAsync(searchService.Object);
+
+            var target = await NuGetSearchServiceReconnector.CreateAsync(serviceBroker.Object, _jtf, CancellationToken.None);
```
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs]: explicit type

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
```
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs]: explicit type

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs [hunk 9]
```diff
@@ -619,42 +634,48 @@ private static void AddToCache(string cacheKey, BitmapSource iconBitmapImage)
...
-                NuGetUIThreadHelper.JoinableTaskFactory
-                    .RunAsync(ReloadPackageVersionsAsync)
-                    .PostOnFailure(nameof(PackageItemViewModel), nameof(ReloadPackageVersionsAsync));
-            }
+                var cancellationToekn = _cancellationTokenSource.Token;
+                IReadOnlyCollection<VersionInfoContextInfo> packageVersions = await GetVersionsAsync();
+
+                // filter package versions based on allowed versions in packages.config
+                packageVersions = packageVersions.Where(v => AllowedVersions.Satisfies(v.Version)).ToList();
+                var result = packageVersions
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs]: please go ahead and change this existing code to an **explicit type**

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs [hunk 10]
```diff
@@ -619,42 +634,48 @@ private static void AddToCache(string cacheKey, BitmapSource iconBitmapImage)
...
+        private async System.Threading.Tasks.Task ReloadPackageVersionsAsync()
         {
-            if (!_backgroundLatestVersionLoader.IsValueCreated)
+            try
             {
-                NuGetUIThreadHelper.JoinableTaskFactory
-                    .RunAsync(ReloadPackageVersionsAsync)
-                    .PostOnFailure(nameof(PackageItemViewModel), nameof(ReloadPackageVersionsAsync));
-            }
+                var cancellationToekn = _cancellationTokenSource.Token;
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs]: typo, explicit type - or, I'd be okay without this variable at all.

**Comment by donnie-msft** [comment]
My testing of package selection and rendering of the Details Pane didn't portray any improvement with this PR. That's probably outside the scope, but I believe it's an issue worth tracking. 
Did you notice any places that weren't, but could be taking advantage of your new behavior?

**Comment by zivkan** [comment]
> My testing of package selection and rendering of the Details Pane didn't portray any improvement with this PR.

Did you use clumsy or fiddler to simulate slower, less reliable networking? If you have a good network connection, you're less likely to have HTTP timeouts as all your requests complete quickly. Having said that, even when I  use those tools to simulator a slow network, it's not very clear that there's an experience benefit, just that the no-longer-needed requests are cancelled correctly.

Additionally, when you used PM UI in the last 30 minutes, if you want to test this PR's changes again, you'll need to run `dotnet nuget locals http-cache --clear`.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs (hunk 2 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs]: The variable needs to be there. If the instance is disposed while the task is pending, then when the continuation runs `_cancellationTokenSource` might already be disposed, so we need the cancellation token to do cancel properly.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs (hunk 3 above)
Comment (zivkan) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs]: .NET doesn't support async constructors, hence why the usual pattern is to have an static async create method. Is it not clear enough?

We might as well use C# 9's `Type name = new();` feature, and completely forbid `var`.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs (hunk 4 above)
Comment (zivkan) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs]: No, it's testing the instance that `ManagedNuGetSearchService` is using, and therefore `NuGetSearchService.Reconnector._service`, is the correct instance. However, since those all all private, the test validates by mocking `INuGetSearchService`, passing that instance through `IServiceBroker`, and validating that calling a method calls the correct instance. 

In other words, it's testing that the wrapping works correctly. The method being called is just coincidental because there isn't a better way to validate other than calling something.

edit: I'll try to come up with a better name. I was just explaining the intent.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs (hunk 3 above)
Comment (donnie-msft) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Utility/NuGetSearchServiceReconnectorTests.cs]: Looks like you went with `NuGetSearchServiceReconnector target` which meets our Coding Guideline # 10.

**Comment by zivkan** [comment]
> Did you notice any places that weren't, but could be taking advantage of your new behavior?

If someone closes PM UI while the updates count is still being calculated, we could cancel that. On the other hand, if they re-open PM UI within 30 minutes, it would be nice to have the json cached so the update check if faster next time.

From my testing, it appears that the details page uses the same `PackageItemViewModel` that infinite scroll list does. I only added `item.Dispose()` to when infinite scroll list is being cleared. However, from my testing it appears that the instance for the details page (right hand side) is also getting disposed, but I'm not 100% sure about that.

Basically, any place we're calling a service (as defined in NuGet.VisualStudio.Internal.Contracts) or NuGet.Protocol, we should think about cancellation. Another option is find all places we use `CancellationToken.None` and change those 😄

**Comment by donnie-msft** [comment]
How did this PR impact faults? I'd assume you tested before/after, and found fault telemetry was no longer being flooded. Does this PR completely reduce noisy faults or is there more we need to track?

Also, did you check the PM UI's reaction to faults after this PR? The UI ends up in a bad state very easily as it is, but just wanted to make sure you (and I will as well) tried navigation & package management actions, and that faults in the middle of those don't lead to a worse UI state than before.

**Comment by donnie-msft** [comment]
Followed-up offline, so now I can answer my own questions...

> How did this PR impact faults?

This PR's motivation was shifted away from https://github.com/NuGet/Client.Engineering/issues/776, so it's not the main priority. My understanding is it may help, but it's hard to repro and measure, so we can 👀 over time.

> Also, did you check the PM UI's reaction to faults after this PR?

I believe zivkan may be doing a couple quick tests to make sure there's no perf regression.

I've created 2 tracking issues for Versions List cancellation and Deprecation cancellation:
https://github.com/NuGet/Home/issues/10811
https://github.com/NuGet/Home/issues/10812

**Review by nkolev92 — APPROVED** [review]
Seems reasonable!

**Comment by zivkan** [comment]
Having a night for the questions and information to brew, I came up with 3 more similar issues that I think will help with UI responsiveness and reduce the risk of HTTP timeouts, espcially fault reporting for requests that don't impact customer experience:

https://github.com/NuGet/Home/issues/10814
https://github.com/NuGet/Home/issues/10815
https://github.com/NuGet/Home/issues/10816

**Review by donnie-msft — APPROVED** [review]
